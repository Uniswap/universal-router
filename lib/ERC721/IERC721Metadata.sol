// SPDX-License-Identifier: MIT
// taken from openzeppelin@3.4.2 and updated the solidity version to fix remapping issues due to different location of the file

pragma solidity ^0.8.0;

import './IERC721.sol';

/**
 * @title ERC-721 Non-Fungible Token Standard, optional metadata extension
 * @dev See https://eips.ethereum.org/EIPS/eip-721
 */
interface IERC721Metadata is IERC721 {
    /**
     * @dev Returns the token collection name.
     */
    function name() external view returns (string memory);

    /**
     * @dev Returns the token collection symbol.
     */
    function symbol() external view returns (string memory);

    /**
     * @dev Returns the Uniform Resource Identifier (URI) for `tokenId` token.
     */
    function tokenURI(uint256 tokenId) external view returns (string memory);
}
