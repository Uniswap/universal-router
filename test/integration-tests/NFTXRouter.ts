import type { Contract } from '@ethersproject/contracts'
import { RouterPlanner, SeaportCommand } from '@uniswap/narwhal-sdk'
import { expect } from './shared/expect'
import { BigNumber } from 'ethers'
import { WeirollRouter } from '../../typechain'
import { abi as ERC721_ABI } from '../../artifacts/solmate/src/tokens/ERC721.sol/ERC721.json'
import snapshotGasCost from '@uniswap/snapshot-gas-cost'

import NFTX_ZAP_ABI from './shared/abis/NFTXZap.json'
import { resetFork } from './shared/mainnetForkHelpers'
import { ALICE_ADDRESS, COVEN_ADDRESS, DEADLINE } from './shared/constants'
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import { expandTo18DecimalsBN } from './shared/helpers'
import hre from 'hardhat'
const { ethers } = hre
import fs from 'fs'

const nftxZapInterface = new ethers.utils.Interface(NFTX_ZAP_ABI)
const COVEN_VAULT = '0xd89b16331f39ab3878daf395052851d3ac8cf3cd'
const COVEN_VAULT_ID = '333'


describe('NFTX', () => {
  let alice: SignerWithAddress
  let weirollRouter: WeirollRouter
  let covenContract: Contract
  let planner: RouterPlanner

  beforeEach(async () => {
    await resetFork()
    await hre.network.provider.request({
      method: 'hardhat_impersonateAccount',
      params: [ALICE_ADDRESS],
    })
    alice = await ethers.getSigner(ALICE_ADDRESS)
    covenContract = new ethers.Contract(COVEN_ADDRESS, ERC721_ABI, alice)
    const weirollRouterFactory = await ethers.getContractFactory('WeirollRouter')
    weirollRouter = (await weirollRouterFactory.deploy(ethers.constants.AddressZero)).connect(alice) as WeirollRouter
    planner = new RouterPlanner()
  })

  afterEach(async () => {
    await resetFork()
  })

  it('completes a buyAndRedeem order', async () => {
    const { order, value } = getOrderParams(seaportOrders[0])
    const params = order.parameters
    const calldata = seaportInterface.encodeFunctionData('fulfillOrder', [order, OPENSEA_CONDUIT_KEY])

    planner.add(SeaportCommand(value.toString(), calldata))
    const { commands, state } = planner.plan()

    const ownerBefore = await covenContract.ownerOf(params.offer[0].identifierOrCriteria)
    const ethBefore = await ethers.provider.getBalance(alice.address)
    const receipt = await (await weirollRouter.execute(DEADLINE, commands, state, { value })).wait()
    const ownerAfter = await covenContract.ownerOf(params.offer[0].identifierOrCriteria)
    const ethAfter = await ethers.provider.getBalance(alice.address)
    const gasSpent = receipt.gasUsed.mul(receipt.effectiveGasPrice)
    const ethDelta = ethBefore.sub(ethAfter)

    expect(ownerBefore.toLowerCase()).to.eq(params.offerer)
    expect(ownerAfter).to.eq(weirollRouter.address)
    expect(ethDelta.sub(gasSpent)).to.eq(value)
  })
})
