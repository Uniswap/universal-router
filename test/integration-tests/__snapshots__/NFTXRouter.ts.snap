// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NFTX gas: buyAndRedeem w/ random selection 1`] = `
Object {
  "calldataByteLength": 708,
  "gasUsed": 510185,
}
`;

exports[`NFTX gas: buyAndRedeem w/ specific selection 1`] = `
Object {
  "calldataByteLength": 772,
  "gasUsed": 497350,
}
`;
